### PR TITLE
py/runtime.c: Fix op bool for subclasses of native types.

### LINKED ISSUE #5677: Truth value test of class deriving from native class can raise TypeError
Don't know if this has always been like this, but it looks like a bug and otherwise it should probably be mentioned as a difference with CPython for instance. Python standard says
```
By default, an object is considered true unless its class defines either a
__bool__() method that returns False or a __len__() method that returns zero,
when called with the object.
```
But when taking a native type which doesn't implement `__bool__` nor `__len__` this happens:
```
import uctypes

class C:
  pass

class B(uctypes.struct):
  def __init__(self):
    super().__init__(1, 'i')

def TestBool(x):
  print(type(x), x, 'compares', True if x else False)

TestBool(C())
TestBool(uctypes.struct(1, 'i'))
TestBool(B())
```

output:

```
<class 'C'> <C object at 3a7f20> compares True
<class 'struct'> <struct ERROR 1> compares True
Traceback (most recent call last):
  File "D:\foo.py", line 43, in <module>
  File "D:\foo.py", line 39, in TestBool
TypeError: unsupported type for __bool__: 'struct'
```

So even though `struct` itself doesn't explicitly implement `__bool__`, MicroPython falls back to default truth testing. But for a derived class it does not. I cannot pinpoint the exact reason, but for the derived class the unary_op call is forwarded by `instance_unary_op` to `mp_unary_op` which raises an exception, whereas in the other cases (class C above and bare uctypes.struct), `mp_obj_is_true` just falls through. So maybe the logic used in `instance_unary_op` is not correct, or `mp_obj_is_true` could try to catch exceptions but that seems far fetched.

The workaround is to manually add something like
```
def __bool__(self):
  return not not super()
```

### PR DESCRIPTION
**Pull request by jimmo**
Fixes #5677

This is +8 bytes on PYBV11.

### CODE REVIEW

**Comment by stinos** [comment]
I also think this is the correct way to fix it, not 100% sure though. Can you add tests?

**Comment by jimmo** [comment]
> I also think this is the correct way to fix it, not 100% sure though. Can you add tests?

Done.

You could also fix this by mandating that all native types have a unary_op with bool implemented, but just doing that for uctypes.struct alone would be about the same code size change. (Interestingly I couldn't find any types in `py/` that would have this issue though).

**Comment by dpgeorge** [comment]
This is a neat fix, but I'm not sure it is strictly correct.

The `MP_UNARY_OP_BOOL` operator is supposed to correspond to `__bool__()`.  What's done in this PR is make `mp_unary_op(MP_UNARY_OP_BOOL, obj)` have the semantics of "truth", ie try `__bool__()` then try `__len__()`.

It's a tricky one.  The proper fix is probably to have a function `mp_unary_op_maybe(op, obj)` which is the same as `mp_unary_op(op, obj)` except that it won't raise if the op is unsupported, and instead just return `MP_OBJ_NULL`.  And then use `mp_unary_op` in `instance_unary_op`.  That might add quite a bit of code size (a new function) and maybe impact performance.

Alternative workaround (as pointed out) is to make `uctypes.struct` type implement `MP_UNARY_OP_BOOL`.

**Comment by jimmo** [comment]
I've pushed an alternative version (only +4 bytes now).

Relative to the previous version, it no longer falls back to `OP_LEN`, but also no longer explicitly checks for `mp_obj_false`, `mp_obj_true`, `mp_obj_none` (i.e. the first few lines of `mp_obj_is_true`).

False and True are fine, as they implement `.unary_op` (which converts to `OP_BOOL`), which will do the right thing.
So the question is -- can `mp_unary_op(OP_BOOL, x)` ever be called on `x=mp_obj_none`? I can't see any occurences -- the only possible cases are instance_unary_op (can't happen on None), and in the vm directly (OP_BOOL isn't in the bytecode).

**Comment by stinos** [comment]
> Alternative workaround (as pointed out) is to make uctypes.struct type implement MP_UNARY_OP_BOOL

This is what I'm doing now for custom types. But if this were chosen as solution maybe `mp_generic_unary_op` should also implement that by default?

**Comment by dpgeorge** [comment]
The current fix (testing for `MP_UNARY_OP_BOOL` and returning `mp_const_true` if so) looks OK to me.  At least it fixes the issue and doesn't impact any other cases (?).

For a core type to test this with (rather than uctypes which doesn't exist on all ports), one can use `type`:
```python
class A(type): pass
a = A("foo", (), {})
print(not a)
print(True if a else False)
```
With this PR that test runs the same under CPython as uPy.

**Comment by jimmo** [comment]
Rebased, updated the test to use `type` instead of `struct`, and added an explainer comment to the change in mp_unary_op.

**Comment by stinos** [comment]
This should fix https://github.com/micropython/micropython/issues/5677 right?

File: tests/basics/unary_op_builtin.py [hunk 1]
```diff
@@ -0,0 +1,7 @@
+# check user instances derived from builtin without unary op handling for bool.
+class A(type):
```
Comment (dpgeorge) [inline tests/basics/unary_op_builtin.py:2]: I think this test should just go in the existing `unary_op.py` file.  There's no longer a reason it needs to be separated (it should be able to run on all uPy ports).

Comment (jimmo) [reply tests/basics/unary_op_builtin.py]: Yep, good point. Updated.

**Comment by dpgeorge** [comment]
Thanks for updating.  Squashed and merged in 4e39ff221abff9d1d7d7fc654da67a89a7543112

**Pull request closed without merging**